rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,387 @@
+Users might want to mock the behavior of a function within one proof harness, and then mock it a different way for another harness, or even use the original function definition.
+It would be overly restrictive to impose the same stub definitions across all proof harnesses.
+A good example of this is compositional reasoning: in some harnesses, we want to prove properties of a particular function (and so want to use its actual implementation), and in other harnesses we want to assume that that function has those properties.
+
+Users will specify stubs by attaching the `#[kani::stub_by(<original>, <replacement>)]` attribute to each harness function.
+The arguments `original` and `replacement` give the names of functions (as string literals), relative to the crate of the harness (*not* relative to the module of the harness).
+The attribute may be specified multiple times per harness, so that multiple (non-conflicting) stub pairings are supported.
+
+For example, this code specifies that the function `mock_random` should be used in place of the function `rand::random` and the function `my_mod::foo` should be used in place of the function `my_mod::bar` for the harness `my_mod::my_harness`:
+
+```rust
+#[cfg(kani)]
+fn mock_random<T: kani::Arbitrary>() -> T {
+    kani::any()
+}
+
+mod my_mod {
+
+    fn foo(x: u32) -> u32 { ... }
+
+    fn bar(x: u32) -> u32 { ... }
+
+    #[cfg(kani)]
+    #[kani::proof]
+    #[kani::stub_by("rand::random", "mock_random")]
+    #[kani::stub_by("my_mod::foo", "my_mod::bar")]
+    fn my_harness() { ... }
+
+}
+```
+
+### Stub sets
+
+As a convenience, we will provide a macro `kani::stub_set` that allows users to specify sets of stubs that can be applied to multiple harnesses:
+
+```rust
+kani::stub_set! {
+    my_io_stubs,
+    stub_by("std::fs::read", "my_read"),
+    stub_by("std::fs::write", "my_write"),
+}
+```
+
+When declaring a harness, users can use the `#[kani::use_stub_set("<stub_set_name>")]` attribute to apply the stub set:
+
+```rust
+#[cfg(kani)]
+#[kani::proof]
+#[kani::use_stub_set("my_io_stubs")]
+fn my_harness() { ... }
+```
+
+A similar mechanism can be used to aggregate stub sets:
+
+```rust
+kani::stub_set!() {
+    all_my_stubs,
+    use_stub_set("my_io_stubs"),
+    use_stub_set("my_other_stubs"),
+}
+```
+
+### Error conditions
+
+Given a set of `original`-`replacement` pairs, Kani will exit with an error if
+
+1. a specified `replacement` stub does not exist;
+2. the user specifies conflicting stubs for the same harness (i.e., if the same `original` function is mapped to multiple `replacement` functions); or
+3. the signature of the `replacement` function is not compatible with the signature of the `original` function.
+
+### Pedagogy
+
+To teach this feature, we will update the documentation with a section on function and method stubbing, including simple examples showing how stubbing can help Kani handle code that currently cannot be verified.
+
+## Detailed design
```

Comment:
We will likely need some changes to the metadata so we can tell `kani-driver` which files were generated.

Reply:
Yes, that makes sense. I will add this.